src/mbgl/style/style_impl.cpp

```diff
@@ -37,6 +38,17 @@ Style::Impl::Impl(Scheduler& scheduler_, FileSource& fileSource_, float pixelRat
 
 Style::Impl::~Impl() = default;
 
+Style::Impl::Impl(const Style::Impl& other)
+    : pixelRatio(other.pixelRatio)
+    , scheduler(other.scheduler)
+    , fileSource(other.fileSource)
+    , spriteLoader(std::make_unique<SpriteLoader>(pixelRatio))
+    , light(std::make_unique<Light>(*other.light))
+    , observer(&nullObserver) {
```

Comment:
What about all the other members? `images`, `sources`, `layers`, etc.?

Copying the collections is going to be tricky. The wrapper classes should be copied, because they have an observer reference, while the immutable instances (and the immutable vector) should be shared.

Ideally, `Style::Impl` is itself held in an `Immutable`, and we move the wrappers, observers, `Scheduler`, `FileSource`, `AsyncRequest`, and `SpriteLoader` to `Style`. Then `MapSnapshotter` can accept an `Immutable<Style::Impl>` and doesn't have to handle making copies of anything, while the dirty work of copying all the wrappers and setting new observers can be centralized in `Style(const Style&)`.